Mathlib/Tactic/Linter/UnusedInstancesInType.lean

```diff
@@ -0,0 +1,222 @@
+    {thm.name.unusedInstancesMsg unusedParams}\n\
+    <extra caption>"
+```
+which logs
+
+> \`{foo}\` has the hypothes[is/es] {p₁}, {p₂}, ..., and {pₙ} which [is/are] not used in the
+  remainder of the type.
+>
+> \<extra caption\>
+>
+> This linter can be disabled with \`set_option {linter.fooLinter.name} false\`
+
+pluralizing as appropriate.
+-/
+@[nolint unusedArguments] -- TODO: we plan to use `_cmd` in future
+def _root_.Lean.Syntax.logUnusedInstancesInTheoremsWhere (_cmd : Syntax)
+    (instanceTypeFilter : Expr → Bool)
+    (log : InfoTree → ConstantVal → Array Parameter → TermElabM Unit)
+    (declFilter : ConstantVal → Bool := fun _ => true) : CommandElabM Unit := do
+  for t in ← getInfoTrees do
+    let thms := t.getTheorems (← getEnv) |>.filter declFilter
+    for thm in thms do
+      thm.onUnusedInstancesWhere instanceTypeFilter fun unusedParams =>
+        -- TODO: restore in order to log on type signature
+        -- t.withDeclSigRef cmd thm.name do
+        log t thm unusedParams
+
+section Decidable
+
+/--
+Checks if `type` is an application of (or forall with return type which is an application of) a
+`Decidable*` constant. Specifically, checks if the constant is one of:
+- `Decidable`
+- `DecidablePred`
+- `DecidableRel`
+- `DecidableEq`
+- `DecidableLE`
+- `DecidableLT`
+
+Ignores metadata and cleans up annotations.
+-/
+def isDecidableVariant (type : Expr) : Bool :=
+  type.isAppOrForallOfConstP fun n =>
+    n == ``Decidable     ||
+    n == ``DecidablePred ||
+    n == ``DecidableRel  ||
+    n == ``DecidableEq   ||
+    n == ``DecidableLE   ||
+    n == ``DecidableLT
+
+/--
+The `unusedDecidableInType` linter checks if a theorem's hypotheses include `Decidable*` instances
+which are not used in the remainder of the type. If so, it suggests removing the instances and using
+`classical` or `open scoped Classical in`, as appropriate, in the theorem's proof instead.
+
+This linter fires only on theorems. (This includes `lemma`s and `instance`s of `Prop` classes.)
+-/
+register_option linter.unusedDecidableInType : Bool := {
+  defValue := false
+  descr := "enable the unused `Decidable*` instance linter, which lints against `Decidable*` \
+    instances in the hypotheses of theorems which are not used in the type, and can therefore be \
+    replaced by a use of `classical` in the proof."
+}
+
+/-- Detects `Decidable*` instance hypotheses in the types of theorems which are not used in the
+remainder of the type, and suggests replacing them with a use of `classical` in the proof or
+`open scoped Classical in` at the term level. -/
+def unusedDecidableInType : Linter where
+  run := /- withSetOptionIn -/ fun cmd => do
+    /- `withSetOptionIn` currently breaks infotree searches, so do a cheap outermost check
+    until this is fixed: [Zulip](https://leanprover.zulipchat.com/#narrow/channel/270676-lean4/topic/bug.3A.20withSetOptionIn.20creates.20context-free.20info.20nodes/with/556896993) -/
```

Comment:
Let's link direct to https://github.com/leanprover/lean4/pull/11313, to make it easier to detect when this is resolved.